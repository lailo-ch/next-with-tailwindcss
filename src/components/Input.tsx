import * as React from 'react'
import clsx from 'clsx'
import { IconType } from 'react-icons/lib/cjs'

export interface Props {
  className?: string | Record<string, unknown>
  startIcon?: IconType
  label?: string
  helpText?: string
  errorMessage?: string
}

const Input: React.FC<Props & React.HTMLProps<HTMLInputElement>> = ({
  className,
  startIcon: StartIcon,
  label,
  helpText,
  errorMessage,
  ...props
}) => {
  return (
    <div className={clsx(className)}>
      {label && (
        <label
          className={clsx(
            'block mb-1 pl-2',
            'text-xs text-gray-600 uppercase',
            'font-light'
          )}
        >
          {label}
        </label>
      )}
      <div
        className={clsx(
          'flex items-center',
          'rounded-sm',
          'border',
          props.disabled && 'border-dashed bg-gray-300 dark:bg-gray-800',
          !errorMessage && [
            'border-gray-400 focus-within:border-gray-600',
            'dark:border-gray-700 dark-focus-within:border-gray-600',
          ],
          errorMessage && 'border-red-600',
          !errorMessage && 'text-black dark:text-white',
          errorMessage && 'text-red-600'
        )}
      >
        {StartIcon && (
          <StartIcon className={clsx('ml-2', errorMessage && 'text-red-600')} />
        )}
        <input
          {...props}
          className={clsx(
            'bg-transparent w-full py-1 px-2',
            props.disabled && 'cursor-not-allowed'
          )}
        />
      </div>
      {helpText && (
        <div className="pl-2 mt-1 text-xs text-gray-600 font-light">
          {helpText}
        </div>
      )}
      {errorMessage && (
        <div className="pl-2 mt-1 text-xs text-red-600 font-light">
          {errorMessage}
        </div>
      )}
    </div>
  )
}

export default Input
