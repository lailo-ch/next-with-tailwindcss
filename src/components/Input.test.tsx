import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { FiMail } from 'react-icons/fi'

import Input from './Input'

describe('Input', () => {
  it('renders with correct label', () => {
    const wrapper = shallow(<Input label="Hello" />)
    expect(wrapper.find('label').first().text()).toEqual('Hello')
  })

  it('renders with value', () => {
    const wrapper = shallow(<Input value="Hello" />)
    expect(wrapper.find('input').props().value).toEqual('Hello')
  })

  it('matches default snapshot', () => {
    const wrapper = shallow(<Input placeholder="email" />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('matches with icon snapshot', () => {
    const wrapper = shallow(<Input placeholder="email" startIcon={FiMail} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('matches disabled snapshot', () => {
    const wrapper = shallow(<Input placeholder="email" disabled />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('matches with error snapshot', () => {
    const wrapper = shallow(<Input placeholder="email" hasError />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('matches custom class names snapshot', () => {
    const wrapper = shallow(<Input className="custom-class" />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('matches with help text', () => {
    const wrapper = shallow(<Input helpText="my help text" />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
