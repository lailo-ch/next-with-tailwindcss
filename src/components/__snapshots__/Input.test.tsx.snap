// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input matches custom class names snapshot 1`] = `
<Fragment>
  <div
    className="flex items-center rounded-sm border border-gray-400 focus-within:border-gray-600 dark:border-gray-700 dark-focus-within:border-gray-600 text-black dark:text-white custom-class"
  >
    <input
      className="bg-transparent w-full py-1 px-2"
    />
  </div>
</Fragment>
`;

exports[`Input matches default snapshot 1`] = `
<Fragment>
  <div
    className="flex items-center rounded-sm border border-gray-400 focus-within:border-gray-600 dark:border-gray-700 dark-focus-within:border-gray-600 text-black dark:text-white"
  >
    <input
      className="bg-transparent w-full py-1 px-2"
      placeholder="email"
    />
  </div>
</Fragment>
`;

exports[`Input matches disabled snapshot 1`] = `
<Fragment>
  <div
    className="flex items-center rounded-sm border border-dashed bg-gray-300 dark:bg-gray-800 border-gray-400 focus-within:border-gray-600 dark:border-gray-700 dark-focus-within:border-gray-600 text-black dark:text-white"
  >
    <input
      className="bg-transparent w-full py-1 px-2 cursor-not-allowed"
      disabled={true}
      placeholder="email"
    />
  </div>
</Fragment>
`;

exports[`Input matches with error snapshot 1`] = `
<Fragment>
  <div
    className="flex items-center rounded-sm border border-red-600 text-red-600"
  >
    <input
      className="bg-transparent w-full py-1 px-2"
      placeholder="email"
    />
  </div>
</Fragment>
`;

exports[`Input matches with icon snapshot 1`] = `
<Fragment>
  <div
    className="flex items-center rounded-sm border border-gray-400 focus-within:border-gray-600 dark:border-gray-700 dark-focus-within:border-gray-600 text-black dark:text-white"
  >
    <FiMail
      className="ml-2"
    />
    <input
      className="bg-transparent w-full py-1 px-2"
      placeholder="email"
    />
  </div>
</Fragment>
`;
